h2o-core/src/main/java/water/ExternalFrameReader.java

```diff
@@ -0,0 +1,163 @@
+
+        AutoBuffer ab = new AutoBuffer().flipForReading().clearForWriting(H2O.MAX_PRIORITY);
+        ab.putInt(chunks[0]._len); // num of rows
+        writeToChannel(ab, sock);
+
+        for (int rowIdx = 0; rowIdx < chunks[0]._len; rowIdx++) { // for each row
+            for(int cidx: selectedColumnIndices){ // go through the chunks
+                ab.flipForReading().clearForWriting(H2O.MAX_PRIORITY); // reuse existing ByteBuffer
+                // write flag weather the row is na or not
+                if (chunks[cidx].isNA(rowIdx)) {
+                    ab.putInt(IS_NA);
+                } else {
+                    ab.putInt(NOT_NA);
+
+                    Chunk chnk = chunks[cidx];
+                    switch (expectedTypes[cidx]) {
+                        case ExternalFrameReader.EXPECTED_INT:
+                            if (chnk.vec().isNumeric() || chnk.vec().isTime()) {
+                                ab.put8(chnk.at8(rowIdx));
+                            } else {
+                                assert chnk.vec().domain() != null && chnk.vec().domain().length != 0;
+                                // in this case the chunk is categorical with integers in the domain
+                                ab.put8(Integer.parseInt(chnk.vec().domain()[(int) chnk.at8(rowIdx)]));
+                            }
+                            break;
+                        case ExternalFrameReader.EXPECTED_DOUBLE:
+                            assert chnk.vec().isNumeric();
+                            if (chnk.vec().isInt()) {
+                                ab.put8(chnk.at8(rowIdx));
+                            } else {
+                                ab.put8d(chnk.atd(rowIdx));
+                            }
+
+                            break;
+                        case ExternalFrameReader.EXPECTED_STRING:
+                            assert chnk.vec().isCategorical() || chnk.vec().isString() || chnk.vec().isUUID();
+                            ab.putStr(getStringFromChunk(chunks, cidx, rowIdx));
+                            break;
+                    }
+
+                }
+                writeToChannel(ab, sock);
+            }
+        }
+        ab.flipForReading().clearForWriting(H2O.MAX_PRIORITY);
+        ab.putInt(ExternalFrameHandler.CONFIRM_READING_DONE);
+        writeToChannel(ab, sock);
+    }
+
+
+
+    private static void writeToChannel(AutoBuffer ab, SocketChannel channel) throws IOException {
+        ab._bb.flip();
+        while (ab._bb.hasRemaining()) {
+            channel.write(ab._bb);
+        }
+    }
+
+    private static String getStringFromChunk(Chunk[] chks, int columnNum, int rowIdx) {
+        if (chks[columnNum].vec().isCategorical()) {
+            return chks[columnNum].vec().domain()[(int) chks[columnNum].at8(rowIdx)];
+        } else if (chks[columnNum].vec().isString()) {
+            BufferedString valStr = new BufferedString();
```

Comment:
Reuse BufferString (should be parameter of the method) - just to avoid allocation.